lib/vnet/dns/osnameservers_darwin.go

```diff
@@ -0,0 +1,101 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+//go:build darwin
+// +build darwin
+
+package dns
+
+import (
+	"bufio"
+	"context"
+	"log/slog"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	confFilePath = "/etc/resolv.conf"
```

Comment:
When I cat this file on my macos system, I see

```conf
#
# macOS Notice
#
# This file is not consulted for DNS hostname resolution, address
# resolution, or the DNS query routing mechanism used by most
# processes on this system.
#
# To view the DNS configuration used by this system, use:
#   scutil --dns
#
# SEE ALSO
#   dns-sd(1), scutil(8)
#
# This file is automatically generated.
#
```

I also think, if we are to use this file or something like it, there's a precedence implied by the order, however we don't appear to respect any precedence

https://github.com/gravitational/teleport/blob/e2f6f104fbba9b86e976843e12b1f26f5f422e61/lib/vnet/dns/dns.go#L259-L261

Reply:
Yeah we shouldn't write to this file because there's no guarantee it would do anything, but MacOS seems to keep it up to date with the current globally configured nameservers, it should be okay to read it, and I'm referencing some other apps that do the same

You're right, there is an implied precedence in the order. We could start adding delays in queries to each successive entry and implement a more sophisticated "happy eyeballs" algorithm, but for now this works. In theory every nameserver is supposed to be able to resolve every name

Reply:
> Yeah we shouldn't write to this file because there's no guarantee it would do anything, but MacOS seems to keep it up to date with the current globally configured nameservers, it should be okay to read it, and I'm referencing some other apps that do the same

Got it, maybe worth a note somewhere noting that reading typically works, and is common practice for MacOS e.g. see <some canonical reference>.

> You're right, there is an implied precedence in the order. We could start adding delays in queries to each successive entry and implement a more sophisticated "happy eyeballs" algorithm, but for now this works. In theory every nameserver is supposed to be able to resolve every name

The scenario I'm imagining is that somebody is within a corporate network environment that relies on the precedence being respected, e.g. their system is set up to ping `<special-corporate-resolver>` as the highest precedence, and that has certain overrides which are meant to take precedence over `<standard-public-resolver>` that's listed second. 

For example, if people within the network lookup "internal.example.com", `<special-corporate-resolver>` sends them to an internal IP address, but outside the network, `<standard-public-resolver>` just respects a wildcard record for the public site when it sees "internal.". In that case it's actually important to the user that we respect their desired precedence.

Reply:
that would be super flaky since this is UDP and they'd have no way to guarantee which nameserver _actually_ got priority. It's also just not a valid way to do DNS, if you're going to have internal DNS then public resolvers should not resolve those domains at all.

but it wouldn't be that hard to successively add a few hundred millisecond delay to each successive nameserver in the list, I added it to the tracker issue under tech debt https://github.com/gravitational/teleport/issues/39330